symengine/functions.cpp

```diff
@@ -3270,6 +3270,161 @@ RCP<const Basic> trigamma(const RCP<const Basic> &x)
+               or (is_a_Number(*arg)
+                   and not down_cast<const Number &>(*arg).is_exact())) {
+        if (not(is_a<Rational>(*arg)
+                and neq(*down_cast<const Rational &>(*arg).get_den(),
+                        *integer(2)))) {
+            return gamma(add(arg, one));
+        }
+    }
+    return make_rcp<const Factorial>(arg);
+}
+
+Binomial::Binomial(const RCP<const Basic> &n, const RCP<const Basic> &k)
+    : TwoArgFunction(n, k)
+{
+    SYMENGINE_ASSIGN_TYPEID()
+    SYMENGINE_ASSERT(is_canonical(n, k))
+}
+
+bool Binomial::is_canonical(const RCP<const Basic> &n,
+                            const RCP<const Basic> &k) const
+{
+    if (is_a_Number(*k) and down_cast<const Number &>(*k).is_zero()) {
+        return false;
+    } else if (is_a_Boolean(*n) or is_a_Boolean(*k) or is_a_Set(*n)
+               or is_a_Set(*k)) {
+        return false;
+    } else if (eq(*n, *k) and not(is_a<NaN>(*n) or is_a<Infty>(*n))) {
+        return false;
+    } else if (is_a_Number(*k) and is_a_Number(*n)
+               and down_cast<const Number &>(*k).is_negative()) {
+        if (eq(*k, *NegInf)) {
+            if (not(eq(*n, *ComplexInf) or eq(*n, *NegInf))) {
+                return false;
+            }
+        } else {
+            return false;
+        }
+    } else if (is_a<NaN>(*n)) {
+        if ((is_a<Integer>(*k)
+             and down_cast<const Number &>(*k).is_positive())) {
+            return false;
+        }
+    } else if (is_a<Infty>(*n)
+               and (is_a<Integer>(*k)
+                    and down_cast<const Number &>(*k).is_positive())) {
+        if (neq(*n, *ComplexInf)) {
+            return false;
+        }
+    } else if (eq(*k, *Inf)
+               and (is_a<Integer>(*n)
+                    and not down_cast<const Number &>(*n).is_negative())) {
+        return false;
+    } else if (is_same_type(*n, *k)) {
+        if (is_a<Integer>(*n)) {
+            return false;
+        }
+        if (is_a<Rational>(*n) or is_a<RealDouble>(*n)) {
+            if (down_cast<const Number &>(*sub(n, k)).is_negative()) {
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
+RCP<const Basic> Binomial::create(const RCP<const Basic> &n,
+                                  const RCP<const Basic> &k) const
+{
+    return binomial(n, k);
+}
+
+RCP<const Basic> binomial(const RCP<const Basic> &n, const RCP<const Basic> &k)
+{
+    if (is_a_Number(*k) and down_cast<const Number &>(*k).is_zero()) {
+        return one;
+    } else if (is_a_Boolean(*n) or is_a_Boolean(*k) or is_a_Set(*n)
+               or is_a_Set(*k)) {
+        throw SymEngineException(
+            "Boolean and Set objects not allowed in this context.");
+    } else if (eq(*n, *k) and not(is_a<NaN>(*n) or is_a<Infty>(*n))) {
+        return one;
+    } else if (is_a<NaN>(*n)) {
+        if ((is_a<Integer>(*k)
+             and down_cast<const Number &>(*k).is_positive())) {
```

Comment:
Why? Just return `NaN` if any one of `n` or `k` is a `NaN`

Reply:
This is `SymPy`'s output:
```
In [5]: binomial(nan, x)
Out[5]: 
⎛nan⎞
⎜   ⎟
⎝ x ⎠

In [6]: binomial(nan, 3/2)
Out[6]: 
⎛nan⎞
⎜   ⎟
⎝1.5⎠

In [7]: binomial(nan, -3)
Out[7]: 
⎛nan⎞
⎜   ⎟
⎝ -3⎠
```

Reply:
If one of the arguments is `NaN`, the value of the function doesn't make sense. Therefore a `NaN` should be returned,